src/misc/umap.c

```diff
@@ -0,0 +1,864 @@
+ * \param distances Pointer to a vector of edge lengths. Similarity graphs for
+ *   UMAP are often originally meant in terms of similarity weights (e.g. correlation between
+ *   high-dimensional vectors) and converted into distances by crude dist := 1 - corr. That is
+ *   fine here too. If this argument is a NULL pointer (NULL), all lengths are assumed equal.
+ * \param layout Pointer to the n x 2 matrix where the layout coordinates will be stored. Only 2D
+ *   embedding are currently supported, as they are by far more common than any higher dimensions.
+ * \param min_dist A fudge parameter that decides how close two unconnected vertices can be in the
+ *   embedding before feeling a repulsive force. It should be positive. Typically, 0.01 is a good
+ *   number.
+ * \param epochs Number of iterations of the main stochastic gradient descent loop on the
+ *   cross-entropy. Usually, 500 epochs can be used if the graph is the graph is small
+ *   (less than 50k edges), 50 epochs are used for larger graphs.
+ * \param sampling_prob The fraction of vertices moved at each iteration of the stochastic gradient
+ *   descent (epoch). At fixed number of epochs, a higher fraction makes the algorithm slower.
+ *   Vice versa, a too low number will converge very slowly, possibly too slowly.
+ *
+ * \return Error code.
+ *
+ */
+igraph_error_t igraph_layout_umap(const igraph_t *graph, const igraph_vector_t *distances,
+        igraph_matrix_t *layout, igraph_real_t min_dist, igraph_integer_t epochs, igraph_real_t sampling_prob) {
+
+
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    /* probabilities of each edge being a real connection */
+    igraph_vector_t umap_weights;
+    /* The smoothing parameters given min_dist */
+    igraph_real_t a, b;
+
+    if (min_dist <= 0) {
+        IGRAPH_ERRORF("Minimum distance should be positive, but found %f.",
+                IGRAPH_EINVAL, min_dist);
+    }
+
+    if (epochs < 0) {
+        IGRAPH_ERRORF("Number of epochs should be non-negative, but found %"
+                IGRAPH_PRId ".",
+                IGRAPH_EINVAL, epochs);
+    }
+
+    if (sampling_prob <= 0) {
+        IGRAPH_ERRORF("Sampling probability should be positive, but found %f.",
+                IGRAPH_EINVAL, sampling_prob);
+    }
+
+    /* UMAP is sometimes used on unweighted graphs, that means distances are always zero */
+    IGRAPH_CHECK(igraph_i_umap_check_distances(distances, no_of_edges));
+
+    RNG_BEGIN();
+    IGRAPH_VECTOR_INIT_FINALLY(&umap_weights, no_of_edges);
+
+    /* Make combined graph with smoothed probabilities */
+    IGRAPH_CHECK(igraph_i_umap_find_prob_graph(graph, distances, &umap_weights));
+
+    /* From now on everything lives in probability space, it does not matter whether
+     * the original graph was weighted/distanced or unweighted */
+
+    /* Skip spectral embedding for now, initialize at random */
+    igraph_layout_random(graph, layout);
```

Comment:
Are the spectral embedding functions that igraph has not suitable here?

BTW those functions should probably have `layout` counterparts. This is a very useful graph layout in its own right.  If we do that first, that will perhaps make it easier to use it in UMAP?

Reply:
@GroteGnoom mentioned our spectral layout is currently buggy and none of us has knowledge + time to debug it, so for now we thought rather then starting with something wrong let's use a random layout. Is that true according to your understanding @szhorvat? If our spectral embedding works, I can add an argument to the public API UMAP function that lets the user choose the initialization. Otherwise, I propose we merge this PR as is and then have a separate PR to bugfix the spectral embedding and include that argument here as well

Reply:
I didn't know it was buggy. I have never used it. @GroteGnoom, if you are aware of specific issues that are not documented yet, can you write a bug report?